bottom_whitespace: Show buttons to navigate to next unread narrow.

amanagr opened the pull request:
This will allow users to easily navigate unread content and
increase use of `n` and `p` hotkeys.
<img width="988" alt="Screenshot 2021-11-13 at 2 45 11 PM" src="https://user-images.githubusercontent.com/25124304/141613254-d88cc3c0-f1de-42fc-85c0-07920cfbd147.png">

timabbott commented on static/templates/bottom_whitespace.hbs:
```diff
@@ -0,0 +1,7 @@
+        <path d="M566.66 527.25c0 33.03-24.23 60.05-53.84 60.05H260.29c-29.61 0-53.84-27.02-53.84-60.05 0-20.22 9.09-38.2 22.93-49.09l134.37-120c2.5-2.14 5.74 1.31 3.94 4.19l-49.29 98.69c-1.38 2.76.41 6.16 3.25 6.16h191.18c29.61 0 53.83 27.03 53.83 60.05zm0-281.39c0 20.22-9.09 38.2-22.93 49.09l-134.37 120c-2.5 2.14-5.74-1.31-3.94-4.19l49.29-98.69c1.38-2.76-.41-6.16-3.25-6.16H260.29c-29.61 0-53.84 ...
+    </svg>
+</div>
+<div id="loading_newer_messages_indicator"></div>
```
Maybe it's worth calling this `message_feed_bottom_whitespace`?

amanagr replied:
Okay

timabbott commented on static/js/click_handlers.js:
```diff
@@ -619,6 +619,21 @@ export function initialize() {
+            e.stopPropagation();
+            e.preventDefault();
+        },
+    );
```
This might be more readable with two similar click handlers rather than doing the `e.target` check.

amanagr replied:
Done.

timabbott commented on static/templates/bottom_whitespace.hbs:
```diff
@@ -0,0 +1,12 @@
+    <a type="button" class="button animated-purple-button next_topic tippy-zulip-tooltip" data-tippy-content="Narrow to an unread topic (n)" data-tippy-placement="bottom">Next unread topic</a>
+    <a type="button" class="button animated-purple-button next_pm tippy-zulip-tooltip" data-tippy-content="Narrow to an unread PM (p)" data-tippy-placement="bottom">Next unread PM</a>
+    <a type="button" href="#recent_topics" class="button animated-purple-button">Recent topics</a>
+</div>
```
Should these have the `animated-purple-button` class?  

Also, should the "Recent topics" button have a tooltip with the `(t)` keyboard shortcut indicated?

amanagr replied:
> Should these have the animated-purple-button class?

Yes, we just change the background color for the button here, rest of the styles are borrowed from it.

> Also, should the "Recent topics" button have a tooltip with the (t) keyboard shortcut indicated?

I wasn't sure of the tooltip text, added it anyway if you want to edit.

alya commented:
A few comments from playing around with this PR:

1) The tippy tooltips are too invasive -- they pop up quickly, are very noticeable, and the user generally doesn't need them. Can we switch to regular tooltips, the way we have for the "New topic" and "New private message" buttons? I think those will be significantly less in the way.

2) It annoying to have the button jump around depending on how many messages are in the topic you just navigated to. You really want it to be consistently in the same place if so that you can keep clicking it without having to move the mouse up and down. Can we figure out a way to place it just above the compose box even if the topic only has a message or two you are looking at?

3) I don't love how the buttons look, though I do appreciate that the aren't as bright as the initial version. Perhaps we should discuss in #design?

4) Currently, in the interleaved stream view the buttons are not shown but the `n` key works. Unless it's too complicated to show the buttons from a technical perspective, this probably merits a #design conversation as well.

Sorry I missed this PR earlier!

alya commented:
I just had the thought that we should make sure something reasonable happens if you're viewing a stream you aren't subscribed to, which has a "Subscribe" button in a similar location.

alya commented:
Also, for visual consistency, we should probably have these buttons look the same/similar to the "Subscribe" button.

timabbott commented:
I suppose we can update this now that we've merged the Spectators version of this feature?

amanagr commented:
Yes.

amanagr commented:
<img width="1469" alt="Screenshot 2022-03-28 at 9 33 53 PM" src="https://user-images.githubusercontent.com/25124304/160592698-c6cf0977-2f4a-4e98-890b-a5a689990566.png">
<img width="1469" alt="Screenshot 2022-03-28 at 9 34 05 PM" src="https://user-images.githubusercontent.com/25124304/160592725-b69aa980-5ce3-4252-87da-48b34026952b.png">
<img width="1469" alt="Screenshot 2022-03-28 at 9 34 18 PM" src="https://user-images.githubusercontent.com/25124304/160592729-896fe086-7656-40b2-92f1-71d4bb420de3.png">
<img width="1469" alt="Screenshot 2022-03-28 at 9 34 25 PM" src="https://user-images.githubusercontent.com/25124304/160592732-57a27fbf-9e93-4cce-9ee1-114857f2b097.png">
<img width="1446" alt="Screenshot 2022-03-29 at 3 41 52 PM" src="https://user-images.githubusercontent.com/25124304/160592753-048216ee-facf-461d-8b6f-2d3b14570a72.png">

amanagr commented:
@alya FYI this is ready for your review after you are free from 5.0 work.

timabbott commented:
I'm test-deploying this on chat.zulip.org.

andersk commented:
The template language is called [Handlebars](https://handlebarsjs.com/), not “handlebar”.

I just saw this button show up at the bottom of the **All messages** view with no unread messages, so it may not be getting updated at all the right times.

timabbott commented:
@amanagr can you update this to address the above feedback and post a summary of its status? I'd like to try getting it deployed on chat.zulip.org again.

amanagr commented:
Rebased.

> I just saw this button show up at the bottom of the All messages view with no unread messages, so it may not be getting updated at all the right times.
Fixed. Thanks!

timabbott commented:
I merged the first commit, and will be deploying this on chat.zulip.org shortly.

timabbott commented:
Removing chat.zulip.org review, since I think the next step for this is to try an alternative design for where the button goes.

One option that would at least be cheap to try is just putting "Next unread topic" in the left sidebar under "Drafts" and see how that works.
